src/main/java/digital/slovensko/autogram/drivers/NativePkcs11SignatureToken.java

```diff
@@ -0,0 +1,162 @@
+package digital.slovensko.autogram.drivers;
+
+import eu.europa.esig.dss.enumerations.SignatureAlgorithm;
+import eu.europa.esig.dss.model.DSSException;
+import eu.europa.esig.dss.model.SignatureValue;
+import eu.europa.esig.dss.model.ToBeSigned;
+import eu.europa.esig.dss.token.DSSPrivateKeyEntry;
+import eu.europa.esig.dss.token.KSPrivateKeyEntry;
+import eu.europa.esig.dss.token.Pkcs11SignatureToken;
+import eu.europa.esig.dss.token.PrefilledPasswordCallback;
+import sun.security.pkcs11.wrapper.CK_ATTRIBUTE;
+import sun.security.pkcs11.wrapper.PKCS11;
+import sun.security.pkcs11.wrapper.PKCS11Constants;
+import sun.security.pkcs11.wrapper.PKCS11Exception;
+
+import java.lang.reflect.Field;
+import java.security.GeneralSecurityException;
+import java.security.PrivateKey;
+import java.security.Signature;
+import java.security.spec.AlgorithmParameterSpec;
+import java.util.Objects;
+
+public class NativePkcs11SignatureToken extends Pkcs11SignatureToken {
+    private static final long CKU_CONTEXT_SPECIFIC = 2L;
+
+    public NativePkcs11SignatureToken(String pkcsPath, PrefilledPasswordCallback prefilledPasswordCallback, int slotIndex) {
+        super(pkcsPath, prefilledPasswordCallback, -1, slotIndex, null);
+    }
+
+    private byte[] sign(final byte[] bytes, final String javaSignatureAlgorithm, final AlgorithmParameterSpec param, final DSSPrivateKeyEntry keyEntry) throws GeneralSecurityException {
+        if (!(keyEntry instanceof KSPrivateKeyEntry)) {
+            throw new IllegalArgumentException("Only KSPrivateKeyEntry are supported");
+        }
+
+        final Signature signature = getSignatureInstance(javaSignatureAlgorithm);
+        if (param != null) {
+            signature.setParameter(param);
+        }
+        var pk = ((KSPrivateKeyEntry) keyEntry).getPrivateKey();
+
+        signature.initSign(pk);
+        runContextSpecificLoginIfNeeded(signature, pk);
+        signature.update(bytes);
+        return signature.sign();
+    }
+
+    private void runContextSpecificLoginIfNeeded(Signature signature, PrivateKey pk) throws GeneralSecurityException {
+        try {
+            // TODO cache & short-circuit
+            var p11 = getP11(signature);
+            var sessionId = getSessionId(signature);
+            var keyID = getKeyID(pk);
+            var attrs = new CK_ATTRIBUTE[]{new CK_ATTRIBUTE(PKCS11Constants.CKA_ALWAYS_AUTHENTICATE)};
+
+            p11.C_GetAttributeValue(sessionId, keyID, attrs);
+
+            if (isAlwaysAuthenticate(attrs)) {
+                var pin = "123456".toCharArray(); // TODO actually get this somehow from password protection
```

Comment:
ja by som tu poslal rovno ten `prefilledPasswordCallback.getPassword()`. Zatiaľ som videl naživo jeden produkčný mandátny certifikát, ktorý mal CKA_ALWAYS_AUTHENTICATE a mal iba jeden pin. Čiže ten istý pin by sa tu posielal a malo by to fungovať (Vieme otestovať u zákazníka). Čo som sa dočítal o kvalifikovaných elektronických pečatiach, tiež majú iba jeden PIN, takže by to malo fungovať rovnako (s týmto sa dvaja ozvali, takže tiež asi vieme u nich otestovať, ak bude build).

Pri eID to vráti null a vtedy ten C_Login otvorí to eID okienko na zadanie PINu, takže nič sa nezmení na doterajšom fungovaní.

Vychádza mi z toho, že teraz nepotrebujeme podporovať zadávanie osobitného pinu pre kľúč. Skúsil by som najprv takto, či to vyrieši elektronické pečate na Gemalto.

Reply:
Rozumiem, no potom musime asi trosku zmenit aj tak logiku "hore" lebo needsPassword je podla kluca, nie podla drivera.